src/inc/conpty-universal.h

```diff
@@ -30,18 +30,161 @@
+// Arguments:
+// - map: The map to populate with the current processes's environment variables.
+// Return Value:
+// - S_OK if we succeeded, or an appropriate HRESULT for failing
+__declspec(noinline) inline
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map)
+{
+    LPWCH currentEnvVars{};
+    auto freeCurrentEnv = wil::scope_exit([&] {
+        if (currentEnvVars)
+        {
+            (void)FreeEnvironmentStringsW(currentEnvVars);
+            currentEnvVars = nullptr;
+        }
+    });
+
+    RETURN_IF_NULL_ALLOC(currentEnvVars = ::GetEnvironmentStringsW());
+
+    // Block is guaranteed to be double-NULL terminated at a minimum.
+    for (wchar_t const* lastCh{ currentEnvVars }; *lastCh != '\0';)
+    {
+        // Copy current entry into temporary map.
+        const size_t cchEntry{ ::wcslen(lastCh) };
+        const std::wstring_view entry{ lastCh, cchEntry };
+
+        // Every entry is of the form "name=value\0".
+        auto pos = entry.find_first_of(L"=", 0, 1);
+        RETURN_HR_IF(E_UNEXPECTED, pos == std::wstring::npos);
+
+        std::wstring name{ lastCh, pos };
+        std::wstring value{ lastCh + pos + 1, cchEntry - pos - 1 };
+
+        // Don't replace entries that already exist.
+        map.try_emplace(std::move(name), std::move(value));
+
+        lastCh += cchEntry + 1; // Each entry is NULL-terminated.
+    }
+
+    return S_OK;
+}
+
+// Function Description:
+// - Creates a new environment block using the provided vector as appropriate
+//   (resizing if needed) based on the provided environment variable map
+//   matching the format of GetEnvironmentStringsW.
+// Arguments:
+// - map: The map to populate the new environment block vector with.
+// - newEnvVars: The vector that will be used to create the new environment block.
+// Return Value:
+// - S_OK if we succeeded, or an appropriate HRESULT for failing
+__declspec(noinline) inline
+HRESULT EnvironmentMapToEnvironmentStringsW(EnvironmentVariableMapW& map, std::vector<wchar_t>& newEnvVars)
+{
+    // Clear environment block before use.
+    constexpr size_t cbChar{ sizeof(decltype(newEnvVars.begin())::value_type) };
+
+    if (!newEnvVars.empty())
+    {
+        ::SecureZeroMemory(newEnvVars.data(), newEnvVars.size() * cbChar);
+    }
+
+    // Resize environment block to fit map.
+    size_t cchEnv{ 2 }; // For the block's double NULL-terminator.
+    for (const auto&[name, value] : map)
+    {
+        // Final form of "name=value\0".
+        cchEnv += name.size() + 1 + value.size() + 1;
+    }
+    newEnvVars.resize(cchEnv);
+
+    // Ensure new block is wiped if we exit due to failure.
+    auto zeroNewEnv = wil::scope_exit([&] {
+        ::SecureZeroMemory(newEnvVars.data(), newEnvVars.size() * cbChar);
+    });
+
+    // Transform each map entry and copy it into the new environment block.
+    LPWCH pEnvVars{ newEnvVars.data() };
+    size_t cbRemaining{ cchEnv * cbChar };
+    for (const auto&[name, value] : map)

```

Comment:
!!!
I completely forgot that C++17 (?) had destructuring assignment. You've made my day.